[#40749] Consistently calculate % Complete in Work Package hierarchies

aaron-contreras opened the pull request:
See: https://community.openproject.org/work_packages/40749

Current status towards 14.0 (Feature freeze on Wednesday April 10)

- Work packages UI:
    - ✅ display unset % Complete as "-" instead of "0%" in activity
- Popover:
    - ✅ work-based mode
    - ✅ Fix jumpiness of popover with the help of Henriette
    - ✅ Fix updating the angular part when receiving a reply from the hotwire part with the help of Oliver
    - ✅ Display the % complete as simple text in popover
    - ✅ Make progress popover work on work package creation form
    - ✅ status-based mode
    - ✅ Extract locale strings from modal (Aaron)
    - ✅ Only display statuses that are valid transitions from the current status (Aaron)
    - ✅ Display migration information on the popover when only % Complete is set ([figma link](https://www.figma.com/file/znrocQvFNLb5jXRXNo7oeD/Progress-estimation?type=design&node-id=1467-18326&mode=design&t=YqkYf8DLuYt6DrHA-4))
    - 🔜 Preview computed values in popover on changes (Aaron / Attila - not launch critical)
    - ✅ See if jumpiness of the popover dialog can be fixed (between display of popover and loading of content, the size expands) (Aaron - use fixed sizes - 500px)
    - 🔜  See if possible to have units displayed in the popover (h, %) (Jens - not launch critical)
    - ✅  When inline creating a work package in table view, after setting a value for work/remaining work via the modal, it doesn't reflect on the display field as the date field does. (Oliver)
    - ✅ Add Learn More link to modal linking to documentation page (Aaron)
- Business logic:
    - ✅ Implement business logic described by specs
    - ✅ Total % Complete computed from total work and total remaining work
    - ✅ Statuses always have a % Complete, 0 by default
    - ✅ Unsetting remaining work while setting work produces an error
    - ✅ Unsetting work while setting remaining work produces an error
    - ✅ Setting Remaining work will set Work to the same value if unset
        - This limits issues with the backlogs module where the remaining work is set without explicitly setting the work. As a side effect, increasing the remaining work when on the backlogs task board can result in an error if it exceeds the work.
    - ✅ Unset % Complete when Work is 0h
- Migration:
    - ✅ Migrate disabled mode to work-based mode as described by spec
    - ✅ Migrate data for work packages in work-based mode
    - ✅ Add journals with changes (System user) with precise cause explanation
    - ✅ Migrate data for work packages in status-based mode
    - ✅ Recompute totals after migration
    - 🔜 Make sure concurrent modification of a work package while the migration job is performing is safe:
      - if a user changes work package data during that job execution, the changes from the job should be in a different journal. The changes from system must be different from the changes from user. updated_at and lock_version should be cached to check it, and then retried if different from the start. maybe lock too. See Jens comments [here](https://github.com/opf/openproject/pull/14768#discussion_r1555912837) and [here](https://github.com/opf/openproject/pull/14768#discussion_r1555918151).
- Administration:
    - ✅ Updating the default % complete of a status should start a job to update all % Complete in instance
    - ✅ Changing from work-based to status-based should start a job to update all % Complete in instance (Christophe)
    - ✅ Create meaningful journal entry for status changes in administration (System user)
    - ✅ Remove "Update work packages with done ratio" button in statuses page in admin
    - ~~Handle status % complete update through~~ [~~statuses API~~](https://www.openproject.org/docs/api/endpoints/statuses/) Since the API does not provide any write-endpoints, an update can not be carried out via the API
    - ✅ Remove "Disabled" mode from Administration UI
    - 🔜 (skippable) Implement the [new Primer UI](https://www.figma.com/file/znrocQvFNLb5jXRXNo7oeD/Progress-estimation?type=design&node-id=1245-33663&mode=design&t=kwtZPf5RHZqf9iUj-4)
- Misc:
    - ✅ Make it impossible to update % Complete by sending emails
    - ✅ No bulk editing of % Complete anymore
    - ✅ Setting status by commit message should not update % Complete anymore

aaron-contreras committed:
Update schema representer specs to mark % complete as read-only

Sets the expectation in our automated suite for the read-only behavior.

aaron-contreras committed:
Mark done_ratio as read-only on the work package contract

aaron-contreras committed:
Remove done_ratio from payload_representer as it's no longer writable

aaron-contreras committed:
Set done_ratio when updating remaining_hours or estimated_hours

* Computes done_ratio on the work package when updating any of its now
two dependent attributes. If any of them are unset, the done_ratio will
also be unset (at 0).

* Add specs covering the behavior of the SetAttributesService

Ensures the possible code-paths for the SetAttributesService:

1- When done_ratio should be computed.
2- When done_ratio should NOT be computed because it's not based
   off of other fields.
3- What the computation should be.

aaron-contreras committed:
Validate remaining_hours doesn't exceed estimated_hours

This scenario is non-sensical. You can't have more remaining work than
work to do.

aaron-contreras committed:
Update computation for `derived_done_ratio`

* The computation now is solely based on a simple % formula between
  `derived_estimated_hours`.

* The order of operations in which derived values are computed needs
  change. `derived_remaining_hours` needs to be computed before
  `derived_done_ratio` as this is a dependent value for the computation

aaron-contreras committed:
Guard against scenario where the parent is having its first child created

When the parent work package hasn't had a child work package yet, its
`derived_estimated_hours` and `derived_remaining_hours`, an error would
be raised as the difference calculated for the `derived_done_ratio`
would be performed by nil values.

In said case where the work package has no current children, ignore the
computation and default to 0 for the `derived_done_ratio`.

aaron-contreras committed:
Re-work expectations on work package create service spec file

Since `done_ratio` is no longer settable, we need to update the specs
to first compute `done_ratio` and subsequently assert based on the
value we'd expect for `done_ratio`.

aaron-contreras committed:
Lint: Use `Date.current` as suggested by Rubocop

aaron-contreras committed:
Add frozen string literal magic comment

aaron-contreras committed:
Register calculation for done_ratio in work package factory

As this is performed by the SetAttributesService, registering it at
the `after_build` stage most closely mimicks this when using factories.

aaron-contreras committed:
Re-work expectations of the update service for work packages

Again, as the caculation method changed, these specs need to be adapted
to the new behavior of done_ratio being read-only.

aaron-contreras committed:
Parse "remaining hours" on incoming emails to derive done ratios

Since `done_ratio` is now a read-only value, we're deprecating being
able to submit this value via the MailHandler service.

Instead, `remaining_hours` is now parseable and `done_ratio` will be
derived from it.

aaron-contreras committed:
Update MailHandler fixtures

No longer set done_ratio but instead derive it from estimated and
remaining hours.

aaron-contreras committed:
Lint: Don't `tap` to yield factory bot record

aaron-contreras committed:
Update query sums test expectations

aaron-contreras committed:
Update docs on accepted attributes for Mail Handling

cbliard commented on app/services/work_packages/set_attributes_service.rb:
```diff
@@ -281,6 +282,35 @@ def update_duration
+    return unless work_package.remaining_hours_changed? || work_package.estimated_hours_changed?
+
+    work_package.done_ratio = if done_ratio_dependent_attribute_unset?
+                                0
```
Eventually, when both work and remaining work are unset, the % Complete will need to be unset too.
That also means we'll have to add a new migration to remove the non-null constraint.

cbliard replied:
done

cbliard commented on spec/services/work_packages/set_attributes_service_spec.rb:
```diff
@@ -145,6 +145,68 @@
     it_behaves_like 'service call'
   end
 
+  context 'for done_ratio' do
```
I find it strange to have the implementation and the spec in 2 different commits. Didn't you write them at the same time?

aaron-contreras replied:
I don't recall why. I probably didn't finish late one afternoon, committed, then continued with missing specs the next day. I'll squash them

cbliard commented on spec/services/work_packages/set_attributes_service_spec.rb:
```diff
@@ -145,6 +145,68 @@
+      context 'when both estimated_hours and remaining_hours are unset' do
+        let(:call_attributes) { { estimated_hours: nil, remaining_hours: nil } }
+        let(:expected_attributes) do
+          { estimated_hours: nil, remaining_hours: nil, done_ratio: 0 }
```
like said in the other comment, expected values will eventually be `{ estimated_hours: nil, remaining_hours: nil, done_ratio: nil }`.
But that's ok, we are moving forward step by step.

cbliard replied:
done

cbliard commented on spec/factories/work_package_factory.rb:
```diff
@@ -125,5 +125,12 @@
                                     updated_at: work_package.journals.maximum(:updated_at))
       end
     end
+
+    callback(:after_build) do |work_package, _evaluator|
+      if work_package.estimated_hours.present? && work_package.remaining_hours.present?
+        work_package.done_ratio = (work_package.estimated_hours - work_package.remaining_hours) \
+                                  / work_package.estimated_hours * 100
+      end
+    end
```
Nice 👍

cbliard commented on docs/installation-and-operations/configuration/incoming-emails/README.md:
```diff
@@ -184,17 +184,18 @@ The subject of the work package that shall be created is derived from the subjec
+| Version         | sets the version                                      | version:v4.1.0                   |
+| Start date      | sets the start date                                   | start date:2015-02-28            |
+| Due date        | sets the finish date                                  |                                  |
+| Estimated hours | sets the estimated hours. Use a number                | Estimated hours:10.5             |
+| Remaining hours | sets the remaining hours. Use a number                | Remaining hours:2.5              |
```
That's where I do not know if we should use the old names or the new names (and have the old ones still working but undocumented). I suppose using the old names does not hurt, and it can still be revisited later.

cbliard replied:
And the table looks nice now that you have added some padding to it.

aaron-contreras replied:
I second guessed on the names too but I thought it could be something to address when we ensure the same terms are used throughout the code base (solely work, remaining_work, percent_complete)

cbliard commented on spec/models/query/results_sums_integration_spec.rb:
```diff
@@ -130,7 +130,7 @@
   end
   let(:type) do
-    create(:type).tap do |t|
+    create(:type) do |t|
```
Is it this kind of things that triggered you to create the Rubocop pull request?

aaron-contreras replied:
Yes :)

cbliard reviewed (commented):
It looks great so far! Nice tests.

aaron-contreras committed:
Compute remaining_hours from other fields when in Status-based mode

Rules when using a status-based mode for keeping % Complete in sync
with the other two work fields apply a bit differently. To ensure
the correct order of operations and in-time updates, this commit moves
the necessary updates to the service layer along with the computation
formulas for correctly deriving fields from one another.

aaron-contreras committed:
Ensure remaining_hours isn't writable on "status-based" mode

aaron-contreras committed:
Re-work UpdateAncestorsService specs

aaron-contreras committed:
Remove ability to bulk update % complete on work packages

This is no longer supported as % Complete is now a read-only value.

aaron-contreras committed:
Interact with progress fields via a modal

* Renders a modal with an edit form for work and remaining work fields.

* Renders a turbo frame within a spot-drop-modal.

* Handles propagating changes from Rails form submissions in an Angular
  context.

cbliard committed:
Merge branch 'dev' into feature/40749-consistent-calculation-of-percent-complete

cbliard committed:
Make done_ratio nil by default

cbliard committed:
Add test scenarios from wp #40749

cbliard committed:
Merge branch 'dev' into feature/40749-consistent-calculation-of-percent-complete

cbliard committed:
Switch to double quotes for files modified on this branch

cbliard committed:
Switch to double quotes for files modified on this branch

cbliard committed:
Merge pull request #14958 from opf/feature/40749--make_done_ratio_nil_by_default

Make done_ratio nil by default

cbliard committed:
Make spec slightly faster

cbliard committed:
Remove useless parent_id merge in WorkPackages::CopyService

It's already part of the writable attributes, so it's already there.

cbliard committed:
Convert to double-quotes

cbliard committed:
Copying a work package copies over the % Complete value

It fixes the feature test ./spec/features/projects/copy_spec.rb:143

cbliard committed:
Fix test: percentageDone is no longer editable

As percentageDone (aka % Complete) is not editable anymore, it does not
make sense anymore to test for it in the shared work package access
spec.

cbliard committed:
Skip this test until we have the progress popover

cbliard committed:
Fix grouping of unset % Complete value

cbliard committed:
Display unset % Complete correctly

Display a '-' instead of misleading '0%' when unset.

cbliard committed:
Fix progress-related feature tests

cbliard committed:
Merge branch 'dev' into feature/40749-consistent-calculation-of-percent-complete

cbliard committed:
Update total % Complete when changing parent work or remaining work

cbliard committed:
Show total work/remaining work if set and children are present

Before that it would have been hidden if value was same as total value. Refs #50959

cbliard committed:
Make total % complete (derived_done_ratio) nil by default

Total % complete is hidden when not set.

cbliard committed:
Unset % complete when changing to a status without a default % complete value

As per specs in https://community.openproject.org/wp/40749, when
changing the status of a work package to a status that does not have a
default % complete value, the % complete value should be unset.

cbliard committed:
Keep progress attributes consistent when updated

Implement business rules as described in
https://community.openproject.org/wp/40749

cbliard committed:
RSpec: display full backtrace of causing exception

I want to debug that 'The PDF export could not be saved: undefined method '>' for nil:NilClass' error that I can not reproduce locally.

cbliard committed:
Merge branch 'dev' into feature/40749-consistent-calculation-of-percent-complete

cbliard committed:
Fix pdf export with nil derived_done_ratio

cbliard committed:
Fix failing test

As setting the status to a status without any % complete now unsets the
remaining work, the test had to be updated.

aaron-contreras committed:
Autofocus the clicked field upon the modal opening

In order to forward context to the modal, it will autofocus the clicked
on field that triggered it to open.

cbliard committed:
Disable failing storages tests from dev branch

Being Pragmatic.

cbliard committed:
Disable failing storages tests from dev branch

Being Pragmatic.

cbliard committed:
Fix test expectations

Now that the remaining work increases by the same amount a work has increased when it's changed, some tests had to be adapted.

cbliard committed:
Add doc to track technical debt

aaron-contreras committed:
Add success toast for saves on progress modal

aaron-contreras committed:
Add status-based version of progress modal

cbliard committed:
Move md file out of spec directory

The CI tries to run it, and produces a syntax error.

cbliard committed:
Keep % complete consistent with work and remaining work

- Avoid recomputing % complete when work and remaining work are invalid
  to avoid having errors on % Complete
- Set work when unset and others are set
- Set remaining work when unset and others are set
- Add more test cases

aaron-contreras committed:
Refresh work package attributes on show page upon successful submit

aaron-contreras committed:
Re-work feature specs with new modal interface

aaron-contreras committed:
Ensure modal closes upon successful save

cbliard committed:
Work or Remaining work cannot be unset if the other is set

This is to ensure that the total work, total remaining work, and total
% Complete are always consistent.

See https://community.openproject.org/wp/40749 for the detailed
specification.

cbliard committed:
Statuses can no longer have a NULL default done ratio

It is 0 by default. This is to ensure that the total work, total
remaining work, and total % Complete are always consistent. Without it,
we could have Work being set and Remaining work being unset, which is
bad for the ancestors totals.

See https://community.openproject.org/wp/40749 for the detailed
specification.

aaron-contreras committed:
Focus clicked input field, field with first error or previous field

This commit ensures an accessible experience to users as dictated
by the Primer guidelines.  The solution for focusing the right field
is built in-house as it's not provided out-of-the-box by Primer's form
fields at the moment.

cbliard committed:
Setting remaining work sets work if unset

cbliard committed:
Round remaining work computed from work and % complete

cbliard committed:
Merge pull request #15159 from opf/feature/40749--tie-work-and-remaining-work-together

Tie work and remaining work together

cbliard committed:
Fix tests about progress modal

Remaining work can only be set if work is set first.

Also refactored `Pages::WorkPackagesTable` a bit to remove duplicated
methods.

cbliard committed:
Fix test features/work_packages/share/access_spec.rb

By refactoring ProgressEditModal to inherit from EditField. The class has been renamed to ProgressEditField as it is about editing a field rather than the modal itself.

cbliard committed:
Correctly format unsetting done_ratio in activity

Before, the `done_ratio` journal formatter was :decimal, which would
always convert nil to 0. And so unsetting the done_ratio would show as
"% Complete changed from 50 to 0".

Now it uses a dedicated :percentage formatter, which will render nil as
expected and also add a "%" sign at the end of the value. And so
unsetting the done_ratio now shows as "% Complete deleted (30%)".

aaron-contreras committed:
Allow interacting with progress modal on new work package form

aaron-contreras committed:
Allow interacting with progress modal on new work package form

cbliard committed:
Can disable pending migration check in dev environment

If OPENPROJECT_DISABLE__MIGRATIONS__CHECK is set to 'true', then the
pending migrations screen will not be displayed. Useful to test some
stuff while working on migrations. But a yellow annoying banner will be
displayed in the bottom of OpenProject page instead.

oliverguenther committed:
Fix duplicate import of TimezoneService

aaron-contreras committed:
Allow status_id being set in ProgressController

cbliard committed:
Migrate progress values for disabled and work-based modes

This is not complete yet:
- The values are migrated for work-based and disabled mode.
- The disabled mode is not allowed anymore (in the setting definition
  allowed values).
- The values still need to be migrated for status-based mode.
- The totals still need to be recomputed.
- The administration UI still needs to have the disabled mode removed,
  and text changed.
- The journals still need to be skipped if nothing has changed.
- The journal `cause` text still needs to be more precise.

aaron-contreras committed:
Disable setting status_id inside the modal while creating a work package

aaron-contreras committed:
Add component specs for modal

* Extracts base class out of both component modes.

aaron-contreras committed:
Add component specs for modal

* Extracts base class out of both component modes.

aaron-contreras committed:
Fix update_ancestors_spec for split view create

* This needs a better solution down the line to deal with the progress
  modal being able to deal with intermediate updates without actually
  persisting on create pages.

cbliard committed:
Merge branch 'dev' into feature/40749-consistent-calculation-of-percent-complete

cbliard committed:
Make FactoryBot wp created_at and updated_at identical

cbliard committed:
Store journal entry for the progress values migration

cbliard committed:
Unset % Complete when work is 0h

When work is 0h, then remaining work is 0h too, but % Complete cannot be
derived, so it is unset.

aaron-contreras committed:
Distinguish field selectors in specs based on create/update context

The selector for the progress edit fields varies based on whether it's
in a create or update context. This commit ensures that this
distinction is made when inferring the `#display_selector` method and
provided by the WorkPackageTable object.

aaron-contreras committed:
Ensure input is disabled while inFlight to prevent race conditions

Fixes custom_actions feature spec

aaron-contreras committed:
Merge branch 'dev' into feature/40749-consistent-calculation-of-percent-complete

aaron-contreras committed:
Set cursor position past the last character on input fields within modal

aaron-contreras committed:
Render the minimum fractional part for numeric fields in progress modal

This ensures that 10.0 is rendered as 10, 5.50 rendered as 5.5 and
5.75 rendered as 5.75.

aaron-contreras committed:
Add feature specs for possible modal scenarios

aaron-contreras committed:
Disable editing values on the progress display fields

* Ensures interactions are solely via the modal.

cbliard committed:
Add status column to table helpers

Works by adding the name of the status in the table.

Raises an error if the status name does not exist

cbliard committed:
Migrate progress values for status-based mode

And amend the WorkPackage factory to force the `done_ratio` different
from the status if it was given to the factory, as `done_ratio` is
automatically set in a before_save hook when in status-based mode.

cbliard committed:
Autocorrect with erb-lint

cbliard committed:
Remove Disabled option from Progress calculation setting

Also renamed the other options to Work-based and Status-based, and added
an explanation of the two options.

lindenthal commented:
Hi @cbliard and @aaron-contreras,

gr**E**aT progr**E**ss!

![grafik](https://github.com/opf/openproject/assets/1019287/0d91a7e5-8f84-4734-9cb2-06d36f1936f7)

In the modal the unit is unclear. It is not a big UX problem since it is shown in the table. And we will address the multi-unit support in 14.1. 

But if you have an idea how to fix this towards the multi-unit approach please let me know.


Best
Niels

cbliard committed:
Compute totals during progress values migration

Using work_package_hierarchies table to join and sum the values in one
request.

cbliard committed:
Always show % Complete column in statuses administration

It has been there since the initial addition of status based mode in
4fe14e71c2d.

Also remove the possibility to set an empty/unset % Complete for
statuses.

cbliard committed:
Autocorrect with erb-lint

cbliard committed:
Make test descriptions more meaningful

cbliard committed:
refactor: remove some useless statements

oliverguenther committed:
Add 14.0 release notes draft for breaking change

ulferts commented on app/workers/work_packages/update_progress_job.rb:
```diff
@@ -0,0 +1,257 @@
+        done_ratio,
+        NULL::double precision AS total_work,
+        NULL::double precision AS total_remaining_work,
+        NULL::integer AS total_p_complete
```
I'd include the `updated_at` or `lock_version` here. Otherwise, a user might loose their input in case they update the work package between the time the `temp_...` table is filled and then written back into the `work_packages` table. When writing back, the sql could bail on rows that have a higher `updated_at`/`lock_version` than the one in the `temp_...` table.

cbliard replied:
Indeed, this is a possibility.

It should then be made more robust.
I'll add it on the TODO.

ulferts commented on app/workers/work_packages/update_progress_job.rb:
```diff
@@ -0,0 +1,257 @@
+    WorkPackage.where(id: updated_work_package_ids).find_each do |work_package|
+      Journals::CreateService.new(work_package, system_user)
+        .call(cause: { type: "system_update", feature: "progress_calculation_changed" })
+    end
```
Here, it could happen (don't know how likely it is), that the job runs, updates the values in the `work _packages` table but hasn't completed the journal creation yet, that a user comes along and updates the work package. The updates to the work/done_ratio would be attributed to the user.

The larger the dataset, the more likely something like this is to happen. That being said, it might be something we are willing to accept. 

A way around that might be to do the updating of the work packages one at a time (potentially in combination with aquiring [advisory locks as done by the services](https://github.com/opf/openproject/blob/dev/app/services/shared/service_context.rb#L44)). Admittedly, that would be slower to execute.

cbliard replied:
That's a valid point. But why would the modification be attributed to the user? The change is in a transaction (one created by the job) and the transaction will only be commited at the end of the job, so if a user changes something, this user will not see the not-yet-commited changes of the job, right?

Maybe I'm wrong and the job does not crete a transaction on its own.
I just checked and it seems the job does not create any transaction. I need to change that then.

Not completely sure of how this could be fully prevented. I'll add it as a TODO.

cbliard committed:
Fix table helpers: do not ignore header if first row value is empty

When using `match_table` helper, some column were missing on the output.
It's because the headers where read from the first row, and the ones
without any value were stripped out. As this information is used to know
how to pull column data from work packages, the data was missing.

ulferts reviewed (commented):
I only took a look at the update job, @cbliard . 

There is an edge case (as described below) but I think we could live with that one being there.

cbliard committed:
refactor: cleaning up spec comments

aaron-contreras committed:
Add feature specs for status-based modal behavior

cbliard committed:
refactor: put methods in the same order as in case statement

aaron-contreras committed:
Fix inline create rendering

* TODO: The values don't seem to be populating on the
  display field after clicking the save button on the modal.

cbliard committed:
Put journal "cause" as first item in journal details

This will displayed it first in the activity views.

cbliard committed:
Fix tests with indifferent access to details hash

oliverguenther committed:
Reload inline-create after changes to changeset state

cbliard committed:
Update explanation text about progress calculation mode

cbliard committed:
Update work packages when status % complete is changed via UI

The % Complete and Remaining work fields are updated in status-based
mode whenever the % complete of a status is changed via the UI.

https://community.openproject.org/wp/40749

ulferts committed:
use WorkPackages::UpdateService to update work packages via changeset

ulferts committed:
Merge remote-tracking branch 'origin/dev' into feature/40749-consistent-calculation-of-percent-complete

cbliard committed:
Use a rubocop version that works

cbliard committed:
Update work packages when changing to status-based mode

https://community.openproject.org/wp/40749

aaron-contreras committed:
Ensure only valid status transitions are allowed in the modal

aaron-contreras committed:
Localize modal strings

aaron-contreras committed:
Render migration warning banner message when % complete is the only field set

aaron-contreras committed:
Set a fixed width for the modal to prevent jumpiness

aaron-contreras committed:
Update modal title

cbliard committed:
Remove "Update work package % Complete values" button in admin

It's not needed anymore are the values are constantly kept in sync.

cbliard committed:
Merge branch 'dev' into feature/40749-consistent-calculation-of-percent-complete

cbliard committed:
Fix frontend build

cbliard committed:
Fix test: remove no longer existing route from tests

klaustopher commented on app/components/work_packages/progress/status_based/modal_body_component.rb:
```diff
@@ -0,0 +1,46 @@
+module WorkPackages
+  module Progress
+    module StatusBased
+      # rubocop:disable OpenProject/AddPreviewForViewComponent
+      class ModalBodyComponent < BaseModalComponent
```
just an FYI: You can also add the disable comment at the end of the line containing the offense. Has the advantage that you don't need to enable it again:

```suggestion
      class ModalBodyComponent < BaseModalComponent # rubocop:disable OpenProject/AddPreviewForViewComponent
```

cbliard replied:
@aaron-contreras this FYI is for you ;)
Thanks! I'll commit it.

aaron-contreras replied:
TIL :)

cbliard committed:
Handle negative work and/or remaining work values

klaustopher commented on app/controllers/statuses_controller.rb:
```diff
@@ -63,6 +63,7 @@ def update
     if @status.update(permitted_params.status)
       flash[:notice] = I18n.t(:notice_successful_update)
       redirect_to action: "index"
+      apply_status_p_complete_change
```
```suggestion
      apply_status_p_complete_change
      redirect_to action: "index"
```

The `render` or `redirect` call should be the last thing in a controller action.

cbliard replied:
You're totally right. I don't know why I did this.

klaustopher commented on app/controllers/work_packages/progress_controller.rb:
```diff
@@ -0,0 +1,138 @@
+  end
+
+  def set_work_package
+    @work_package = WorkPackage.find(params[:work_package_id])
```
Eventhough the services validate the permissions via the contracts, we should make sure that we always only allow access to `visible` resources.

```suggestion
    @work_package = WorkPackage.visible.find(params[:work_package_id])
```

cbliard replied:
Good catch 👍

cbliard replied:
@aaron-contreras

cbliard committed:
Reuse user value if field is in error state

klaustopher commented on app/models/changeset.rb:
```diff
@@ -231,15 +231,18 @@ def fix_work_package(work_package)
-                                changeset: self, issue: work_package)
-    if !work_package.save(validate: false) && logger
+    call = WorkPackages::UpdateService
+           .new(model: work_package,
+                user: user || User.anonymous)
+           .call(status:,
+                 journal_notes: I18n.t(:text_status_changed_by_changeset,
+                                       value: text_tag,
+                                       locale: Setting.default_language))
```
Not for right now, but we should maybe add a ticket that we replace those notes by a cause so we can properly show a translated message

cbliard replied:
@ulferts could you address this one?

klaustopher commented on app/services/work_packages/copy_service.rb:
```diff
@@ -77,8 +77,7 @@ def copied_attributes(work_package, override)
     attributes = work_package
                    .attributes
                    .slice(*writable_work_package_attributes(work_package))
-                   .merge("parent_id" => work_package.parent_id,
-                          "custom_field_values" => work_package.custom_value_attributes)
```
`parent_id` is removed here because it is already part of the sliced params?

```ruby
WorkPackages::CopyContract.new(WorkPackage.new, User.first).writable_attributes.include?("parent_id")
#=> true
```

cbliard replied:
Yes, that's something we realized with Jens in a pairing session.
We did not understand why this was here as it seemed it would be picked by the `slice` call above. So we looked at tests, tried with, and without, and as expected the parent is carried over, so we removed it.

oliverguenther committed:
Correct change detection cycle for adding a widget

klaustopher commented on app/workers/work_packages/update_progress_job.rb:
```diff
@@ -0,0 +1,257 @@
+        derive_unset_remaining_work_from_work_and_p_complete
+        derive_unset_work_from_remaining_work_and_p_complete
+        derive_p_complete_from_work_and_remaining_work
+      end
+      if current_mode == "status"
```
```suggestion
      elsif current_mode == "status"
```

I don't think we change `current_mode` in the first if block, so the second block can never be executed. This way we will not even need to look at the second if block.

cbliard replied:
You're right. I used a case statement instead. Reads better IMO.
8c5c2094919

klaustopher commented on lib_static/plugins/acts_as_journalized/lib/journal_changes.rb:
```diff
@@ -31,70 +31,88 @@ def get_changes
-        )
-      )
-    end
+    changes = [
+      get_cause_changes,
+      get_data_changes,
+      get_attachments_changes,
+      get_custom_fields_changes,
+      get_file_links_changes,
+      get_agenda_items_changes
+    ].compact
```
Very good: 👏 👏

klaustopher commented on progress-technical-debt.md:
```diff
@@ -0,0 +1,14 @@
+  * "Work must be >= 0": one we created ourself in https://github.com/opf/openproject/commit/15f52bcdc6cc50c30e14c213133cab4d59715347
+  * "Remaining work must be greater than or equal to 0" is the standard rails message (builtin. no translation needed)
+* In `app/services/work_packages/update_ancestors_service.rb`, `WorkPackage::UpdateAncestorsService#derive_attributes` is updating both `derived_estimated_hours` and `derived_remaining_hours` even when only one needs to be updated. This is because the tests depend on this behavior. (not sure that's worth modifying)
+*
```
I think these should be Code Maintenance tickets

cbliard replied:
yes, you're right. I'll create one.

cbliard replied:
ticket created: https://community.openproject.org/projects/openproject/work_packages/54050/activity

dombesz committed:
Refresh remaining work calculations on the work estimate modal.

aaron-contreras committed:
Ensure user is allowed to see the requested work package

Thanks Klaus!

Co-authored-by: Klaus Zanders <klaustopher@users.noreply.github.com>

cbliard committed:
refactor: Put rubocop disable pragma on the offending line

This avoid having to reenable it later

Co-authored-by: Klaus Zanders <klaustopher@users.noreply.github.com>

cbliard committed:
Make render the final action statement

Co-authored-by: Klaus Zanders <klaustopher@users.noreply.github.com>

cbliard committed:
refactor: handle different progress modes with a case statement

Co-authored-by: Klaus Zanders <klaustopher@users.noreply.github.com>

aaron-contreras commented on app/components/work_packages/progress/status_based/modal_body_component.html.erb:
```diff
@@ -0,0 +1,36 @@
+      <% modal_body.with_row(mt: 3) do |_tooltip| %>
+        <%= render(Primer::Beta::Text.new(font_weight: :semibold)) { t("work_package.progress.label_note") } %>
+        <%= render(Primer::Beta::Text.new) { t("work_package.progress.modal.status_based_help_text") } %>
+        <%= render(Primer::Beta::Link.new(href: "#")) { t(:label_learn_more) } %>
```
Do we have a URL already for this?

cbliard replied:
@aaron-contreras as suggested by Oliver, let's remove the links, and add them back once the documentation path is clear.

cbliard replied:
@aaron-contreras as suggested by Parimal, it should link to https://www.openproject.org/docs/user-guide/time-and-costs/progress-tracking/.

aaron-contreras replied:
Done!

aaron-contreras commented on app/components/work_packages/progress/work_based/modal_body_component.html.erb:
```diff
@@ -0,0 +1,36 @@
+      <% modal_body.with_row(mt: 3) do |_tooltip| %>
+        <%= render(Primer::Beta::Text.new(font_weight: :semibold)) { t("work_package.progress.label_note") } %>
+        <%= render(Primer::Beta::Text.new) { t("work_package.progress.modal.work_based_help_text") } %>
+        <%= render(Primer::Beta::Link.new(href: "#")) { t(:label_learn_more) } %>
```
Same for this

cbliard replied:
Same, https://www.openproject.org/docs/user-guide/time-and-costs/progress-tracking/

aaron-contreras replied:
Done!

dombesz committed:
Preserve cursor position of the progress modal by using morphing on the frame re-rendering.

cbliard committed:
Replace technical debt doc with a code maintenance work package

Code maintenance ticket is https://community.openproject.org/wp/54050.

cbliard committed:
Merge branch 'dev' into feature/40749-consistent-calculation-of-percent-complete

aaron-contreras committed:
Link to the documentation for more info on the inner workings of progress

dombesz committed:
Revert "Preserve cursor position of the progress modal by using morphing on the frame re-rendering."

This reverts commit 714057ff636da70577191514c5825036778c3fa7.

dombesz committed:
Revert "Refresh remaining work calculations on the work estimate modal."

This reverts commit d0232cd90d7ac8eae63dbb71cb06fa7d4f29eafd.

cbliard committed:
Use 'field' mode when setting does not exist in database yet

ulferts committed:
remove done todo

cbliard committed:
Enable remaining work field when there are errors in wp

ulferts reviewed (approved):
Admittedly I did not do a fully fledged review. But the reading and the tests I did all looked positive.

So from my point of view, this can be merged. Thanks @cbliard and @aaron-contreras for bringing this alive.
